### PR TITLE
schedule: simplify the scatter api code

### PR DESCRIPTION
**Pull request by Yisaer**
Signed-off-by: Song Gao <disxiaofei@163.com>

### What is changed and how it works?

Remove some duplicated code in the scatter api handler and expose some useful function for the region scatter.

### CODE REVIEW

File: server/handler.go [hunk 1]
```diff
@@ -786,35 +786,23 @@ func (h *Handler) AddScatterRegionsOperators(regionIDs []uint64, startRawKey, en
...
-	}
 	// If there existed any operator failed to be added into Operator Controller, add its regions into unProcessedRegions
 	for _, op := range ops {
 		if ok := c.GetOperatorController().AddOperator(op); !ok {
-			failureRegionID = append(failureRegionID, fmt.Sprintf("%v", op.RegionID()))
+			failures[op.RegionID()] = fmt.Errorf("region %v failed to add operator", op.RegionID())
 		}
 	}
-	return 100 - (len(failureRegionID) * 100 / len(regions)), errors.New("unprocessed regions:[" + strings.Join(failureRegionID, ",") + "]")
+	return 100 - 100*len(failures)/(len(ops)+len(failures)), nil
```
Comment (rleungx) [inline server/handler.go:805]: Will (len(ops)+len(failures) be 0?

File: server/api/region.go [hunk 2]
```diff
@@ -662,66 +662,53 @@ func (h *regionsHandler) ScatterRegions(w http.ResponseWriter, r *http.Request)
...
 			h.rd.JSON(w, http.StatusBadRequest, err.Error())
 			return
 		}
-		regions := rc.ScanRegions(startKey, endKey, -1)
-		regionMap = make(map[uint64]*core.RegionInfo, len(regions))
-		for _, region := range regions {
-			regionMap[region.GetID()] = region
+		ops, failures, err = rc.GetRegionScatter().ScatterRegionsByRange(startKey, endKey, group, retryLimit)
+		if err != nil {
+			h.rd.JSON(w, http.StatusBadRequest, err.Error())
```
Comment (rleungx) [inline server/api/region.go:691]: Should the error be StatusBadRequest?

File: server/handler.go (hunk 1 above)
Comment (Yisaer) [reply server/handler.go]: updated.

File: server/api/region.go (hunk 2 above)
Comment (Yisaer) [reply server/api/region.go]: replaced as IntervalServerError

File: server/api/region.go [hunk 3]
```diff
@@ -662,66 +662,57 @@ func (h *regionsHandler) ScatterRegions(w http.ResponseWriter, r *http.Request)
...
+			h.rd.JSON(w, http.StatusInternalServerError, err.Error())
+			return
 		}
-		regionsCount = len(regionMap)
 	} else {
 		regionsID := input["regions_id"].([]uint64)
-		regionMap = make(map[uint64]*core.RegionInfo, len(regionsID))
-		for _, id := range regionsID {
-			regionMap[id] = rc.GetRegion(id)
+		ops, failures, err = rc.GetRegionScatter().ScatterRegionsByID(regionsID, group, retryLimit)
```
Comment (nolouch) [inline server/api/region.go:696]: Is this result calculated after the operator finished?

Comment (Yisaer) [reply server/api/region.go:696]: The result (like finished percentage) would be calculated in the handler layer.

**Review by rleungx — APPROVED** [review]
LGTM

**Pull request merged**